src/mon/ElectionLogic.h

```diff
@@ -0,0 +1,349 @@
+   * election epoch. This resets an election (either on local loss or victory,
+   * or when trying a new election round) and the ElectionOwner
+   * should reset any tracking of its own to match. (The ElectionLogic
+   * will further trigger sending election messages if that is
+   * appropriate.)
+   */
+  virtual void notify_bump_epoch() = 0;
+  /**
+   * Notify the ElectionOwner we must start a new election.
+   */
+  virtual void trigger_new_election() = 0;
+  /**
+   * Retrieve this Paxos instance's rank.
+   */
+  virtual int get_my_rank() = 0;
+  /**
+   * Send a PROPOSE message to all our peers. This happens when
+   * we have started a new election (which may mean attempting to
+   * override a current one).
+   *
+   * @param e The election epoch of our proposal.
+   */
+  virtual void propose_to_peers(epoch_t e) = 0;
+  /**
+   * The election has failed and we aren't sure what the state of the
+   * quorum is, so reset the entire system as if from scratch.
+   */
+  virtual void reset_election() = 0;
+  /**
+   * Ask the ElectionOwner if we-the-Monitor have ever participated in the
+   * quorum (including across process restarts!).
+   *
+   * @returns true if we have participated, false otherwise
+   */
+  virtual bool ever_participated() = 0;
+  /**
+   * Ask the ElectionOwner for the size of the Paxos set. This includes
+   * those monitors which may not be in the current quorum!
+   */
+  virtual unsigned paxos_size() = 0;
+  /**
+   * Tell the ElectionOwner we have started a new election.
+   *
+   * The ElectionOwner is responsible for timing out the election (by invoking
+   * end_election_period()) if it takes too long (as defined by the ElectionOwner).
+   * This function is the opportunity to do that and to clean up any other external 
+   * election state it may be maintaining.
+   */
+  virtual void _start() = 0;
+  /**
+   * Tell the ElectionOwner to defer to the identified peer. Tell that peer
+   * we have deferred to it.
+   *
+   * @post  we sent an ack message to @p who
+   */
+  virtual void _defer_to(int who) = 0;
+  /**
+   * We have won an election, so have the ElectionOwner message that to
+   * our new quorum!
+   *
+   * @param quorum The ranks of our peers which deferred to us and
+   *        must be told of our victory
+   */
+  virtual void message_victory(const set<int>& quorum) = 0;
+  /**
+   * Query the ElectionOwner about if a given rank is in the
+   * currently active quorum.
+   * @param rank the Paxos rank whose status we are checking
+   * @returns true if the rank is in our current quorum, false otherwise.
+   */
+  virtual bool is_current_member(int rank) = 0;
+  virtual ~ElectionOwner() {}
+};
+
+/**
+ * This class maintains local state for running an election
+ * between Paxos instances. It receives input requests
+ * and calls back out to its ElectionOwner to do persistence
+ * and message other entities.
+ */
+
+class ElectionLogic {
+  ElectionOwner *elector;
```

Comment:
might want to put the member variables at the end of class . see https://google.github.io/styleguide/cppguide.html#Declaration_Order

Reply:
also, if we will never have a `nullptr` elector, probably we could use a reference instead of a pointer?

Reply:
Is there any reason to make it a reference instead of a pointer for interfaces like this?

Reply:
@gregsfortytwo i think the only reason is readability. as reference enforces more limitation than pointer does.